bittide-instances/exe/post-vex-riscv-tcp-test/Main.hs

```diff
@@ -0,0 +1,177 @@
+  putStrLn "Starting Gdb"
+  withAnnotatedGdbScriptPath gdbScriptPath $ \gdbProgPath -> do
+    let
+      openOcdProc = (proc startOpenOcdPath []){std_err = CreatePipe}
+      gdbProc = (proc "gdb" ["--command", gdbProgPath]){std_out = CreatePipe, std_err = CreatePipe}
+      picocomProc = (proc startPicocomPath [uartDev]){std_out = CreatePipe, std_in = CreatePipe}
+
+      -- Wait until we see "Halting processor", fail if we see an error
+      waitForHalt s
+        | "Error:" `isPrefixOf` s = Stop (Error ("Found error in OpenOCD output: " <> s))
+        | "Halting processor" `isPrefixOf` s = Stop Ok
+        | otherwise = Continue
+
+    putStrLn "Starting OpenOcd..."
+    withCreateProcess openOcdProc $ \_ _ (fromJust -> openOcdStdErr) _ -> do
+      hSetBuffering openOcdStdErr LineBuffering
+
+      putStr "Waiting for halt..."
+      expectLine openOcdStdErr waitForHalt
+      putStrLn " Done"
+
+      putStrLn "Starting Picocom..."
+      withCreateProcess picocomProc $ \maybePicocomStdIn maybePicocomStdOut maybePicocomStdErr _ -> do
+        let
+          picocomStdInHandle = fromJust maybePicocomStdIn
+          picocomStdOutHandle = fromJust maybePicocomStdOut
+
+          -- Create function to log the output of the processes
+          loggingSequence = do
+            threadDelay 1_000_000 -- Wait 1 second for data loggers to catch up
+            putStrLn "Picocom stdout"
+            picocomOut <- readRemainingChars picocomStdOutHandle
+            putStrLn picocomOut
+            case maybePicocomStdErr of
+              Nothing -> pure ()
+              Just h -> do
+                putStrLn "Picocom StdErr"
+                readRemainingChars h >>= putStrLn
+
+          tryWithTimeout :: String -> Int -> IO a -> IO a
+          tryWithTimeout msg dur action = do
+            result <- timeout dur action
+            case result of
+              Nothing -> do
+                loggingSequence
+                assertFailure msg
+              Just r -> pure r
+
+        hSetBuffering picocomStdInHandle LineBuffering
+        hSetBuffering picocomStdOutHandle LineBuffering
+
+        putStrLn "Waiting for Picocom to be ready..."
+        _ <-
+          tryWithTimeout "Timeout waiting for Picocom to be ready" 10_000_000 $
+            waitForLine picocomStdOutHandle "Terminal ready"
+
+        putStrLn "Starting GDB..."
+        withCreateProcess gdbProc $ \_ (fromJust -> gdbStdOut) _ _ -> do
+          hSetBuffering gdbStdOut LineBuffering
+
+          putStrLn "Waiting for \"Starting TCP Client\""
+          _ <-
+            tryWithTimeout "Timeout waiting for \"Starting TCP Client\"" 10_000_000 $
+              waitForLine picocomStdOutHandle "Starting TCP Client"
+
+          let numberOfClients = 1
+          putStrLn $ "Waiting for " <> show numberOfClients <> " clients to connect to TCP server."
+          clients <-
+            tryWithTimeout "Timeout waiting for clients to connect" 10_000_000 $
```

Comment:
You could put `"Timeout waiting for clients "` part inside `tryWithTimeout` so you don't have to repeat it